BoogieLang/Ast_Cfg_Transformation.thy

```diff
@@ -0,0 +1,2280 @@
+      using disj_a by blast
+    thus ?thesis
+    proof -
+      obtain l2_conc where conc_trace: "(A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb, cont0, Normal ns1'') -n\<longrightarrow>^l2_conc (reached_bb, reached_cont, reached_state))" and 
+                           succ_rel: "(j' = Suc l2_conc)"
+        using \<open>\<exists>l2. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb, cont0, Normal ns1'') -n\<longrightarrow>^l2 (reached_bb, reached_cont, reached_state)) \<and> j' = Suc l2\<close> by blast
+      show ?thesis
+        apply (rule assms(7))
+           apply (rule succ_rel)
+          apply (rule conc_trace)
+         apply (rule assms(4))
+          apply (simp)
+         apply (rule assms(5))
+          apply assumption+
+        using assms(6) false_equals_not_true 
+        by blast
+    qed
+  qed
+qed
+
+text \<open>Pair 2: The starting configuration represents a point in the program after a loop and the continuation needs to be adjusted and subsequently entered.
+              (This could be replaced by a simpler lemma.)\<close>
+lemma endblock_skip2:
+  assumes "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb0, KEndBlock (KSeq bb_next cont0), Normal ns3) -n\<longrightarrow>^l (reached_bb, reached_cont, reached_state)"
+      and "bb0 = BigBlock None [] None None" 
+  shows "(valid_configuration A \<Lambda> \<Gamma> \<Omega> posts reached_bb reached_cont reached_state) \<or> 
+            (\<exists> l2. (A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(bb_next, cont0, Normal ns3) -n\<longrightarrow>^l2 (reached_bb, reached_cont, reached_state)) \<and> (l = Suc (Suc l2)) )"
+  using assms
+proof (cases l)
+  case 0
+  then show ?thesis by (metis Ast.valid_configuration_def assms(1) get_state.simps is_final.simps(6) relpowp_fun_conv state.simps(3))
+next
+  case 1: (Suc l1)
+  then show ?thesis 
+  proof (cases l1)
+    case 0
+    from 1 assms this have "(red_bigblock A M \<Lambda> \<Gamma> \<Omega> T (BigBlock None [] None None, KEndBlock (KSeq bb_next cont0), Normal ns3) (reached_bb, reached_cont, reached_state))"
+      by fastforce
+    then show ?thesis
+    proof cases
+      case RedSkipEndBlock thus ?thesis by (simp add: Ast.valid_configuration_def)
+    qed auto
+  next
+    case 2: (Suc l2)
+    from 2 1 have "l = Suc (Suc l2)" by auto
+    from 2 1 assms obtain inter_bb inter_cont inter_state inter_bb2 inter_cont2 inter_state2 where
+      step1: "(red_bigblock A M \<Lambda> \<Gamma> \<Omega> T (BigBlock None [] None None, KEndBlock (KSeq bb_next cont0), Normal ns3) (inter_bb, inter_cont, inter_state))" and 
+      step2: "(red_bigblock A M \<Lambda> \<Gamma> \<Omega> T (inter_bb, inter_cont, inter_state) (inter_bb2, inter_cont2, inter_state2))" and
+      rest: "A,M,\<Lambda>,\<Gamma>,\<Omega>,T \<turnstile>(inter_bb2, inter_cont2, inter_state2) -n\<longrightarrow>^l2 (reached_bb, reached_cont, reached_state)" 
+      by (metis (no_types, opaque_lifting) prod_cases3 relpowp_Suc_D2)
```

Comment:
replace by "by (metis (no_types) etc...)", otherwise the proofs fail on Windows.